flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaDeserializationSchema.java

```diff
@@ -71,16 +88,45 @@ public RowData deserialize(ConsumerRecord<byte[], byte[]> record) throws Excepti
 		throw new IllegalStateException("A collector is required for deserializing.");
 	}
 
+	/**
+	 * There are 4 situations:
+	 * 1. value data only: collector will ignore the deserialization error;
+	 * 2. value data and metadata only:
+	 *   if value fails to deserialize, value decoding format will ignore data;
+	 *   else OutputProjectionCollector will emitRow with key is null;
+	 * 3. key data and value data (no tombstone):
+	 *   if key fails to deserialize and value succeed: OutputProjection ignores the value data;
+	 *   else if key succeed and value fails, value decoding format will ignore;
+	 *   else if both succeed, OutputProjection will emitRow with the key and data.
+	 * 4.key data and value data (tombstone):
+	 *   if key fails, OutputProjection ignores value data;
+	 *   else emitRow with key and null value.
+	 */
 	@Override
 	public void deserialize(ConsumerRecord<byte[], byte[]> record, Collector<RowData> collector) throws Exception {
-		// shortcut if no metadata is required
-		if (!hasMetadata) {
+		// shortcut in case no output projection is required,
+		// also not for a cartesian product with the keys
+		if (keyDeserialization == null && !hasMetadata) {
 			valueDeserialization.deserialize(record.value(), collector);
+			return;
+		}
+
+		// buffer key(s)
+		if (keyDeserialization != null) {
+			keyDeserialization.deserialize(record.key(), keyCollector);
+		}
+
+		// project output while emitting values
+		outputCollector.inputRecord = record;
+		outputCollector.physicalKeyRows = keyCollector.buffer;
+		outputCollector.outputCollector = collector;
+		if (record.value() != null) {
+			valueDeserialization.deserialize(record.value(), outputCollector);
 		} else {
-			metadataAppendingCollector.inputRecord = record;
-			metadataAppendingCollector.outputCollector = collector;
-			valueDeserialization.deserialize(record.value(), metadataAppendingCollector);
+			// trigger collect by hand
+			outputCollector.collect(null);
```

Comment:
We should add a flag, e.g. `upsertMode=true` to enable this behavior. Because for normal kafka connector, the tombstone messages should be skipped.